src/Features/Core/Portable/EditAndContinue/EditAndContinueDiagnosticUpdateSource.cs

```diff
@@ -48,15 +58,34 @@ public ValueTask<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(Workspace w
         /// We do not track the particular reported diagnostics here since we can just clear all of them at once.
         /// </summary>
         public void ClearDiagnostics()
-            => DiagnosticsCleared?.Invoke(this, EventArgs.Empty);
+        {
+            // If ClearDiagnostics is called and there weren't any diagnostics previously, then there is no point incrementing
+            // our version number and potentially invalidating caches unnecessarily.
+            if (_previouslyHadDiagnostics)
+            {
+                _previouslyHadDiagnostics = false;
+                _diagnosticsVersion++;
+            }
+
+            DiagnosticsCleared?.Invoke(this, EventArgs.Empty);
+        }
 
         /// <summary>
         /// Reports given set of project or solution level diagnostics. 
         /// </summary>
-        public void ReportDiagnostics(Workspace workspace, Solution solution, ImmutableArray<DiagnosticData> diagnostics)
+        public void ReportDiagnostics(Workspace workspace, Solution solution, ImmutableArray<DiagnosticData> diagnostics, ImmutableArray<(DocumentId, ImmutableArray<RudeEditDiagnostic> Diagnostics)> rudeEdits)
         {
             RoslynDebug.Assert(solution != null);
 
+            // Even though we only report diagnostics, and not rude edits, we still need to
+            // ensure that the presence of rude edits are considered when we decide to update
+            // our version number.
+            if (diagnostics.Any() || rudeEdits.Any(d => d.Diagnostics.Any()))
```

Comment:
>Diagnostics [](http://example.com/codeflow?start=58&length=11)

The array in rudeEdits won't ever be empty for a given document. So you can just check if the outer array is empty.